include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,385 @@
+/**
+ * @file nsga2_impl.hpp
+ * @author Sayan Goswami
+ *
+ * Implementation of the NSGA-II algorithm. Used for multi-objective
+ * optimisation problems on arbitrary functions.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more Information.
+ */
+
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+#define ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+
+#include "nsga2.hpp"
+#include <set>
+#include <vector>
+#include <map>
+
+namespace ens {
+
+inline NSGA2::NSGA2(const size_t populationSize,
+                    const size_t maxGenerations,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const double epsilon,
+                    const arma::vec lowerBound,
+                    const arma::vec upperBound) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon),
+    lowerBound(lowerBound),
+    upperBound(upperBound)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+std::vector<MatType> NSGA2::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                     MatType& iterate,
+                                     CallbackTypes&&... callbacks)
+{
+  // Make sure for evolution to work at least four candidates are present.
+  if (populationSize < 4 && populationSize % 4 != 0)
+  {
+    throw std::logic_error("NSGA2::Optimize(): population size should be at"
+        " least 4, and, a multiple of 4!");
+  }
+
+  // TODO: Does not work presently "implicit instantiation of undefined template"
+  numObjectives = std::tuple_size<decltype(objectives)>::value;
```

Comment:
This does not work. 😭 

I checked [quite](https://en.cppreference.com/w/cpp/utility/tuple/tuple_size) [a](https://www.murrayc.com/permalink/2015/12/05/modern-c-variadic-template-parameters-and-tuples/) [few](http://www.cplusplus.com/reference/tuple/tuple_size/) places but I haven't been able to resolve it.

Reply:
`sizeof...()` solved it, thanks Marcus!